tests/ObjectReadStateMachineTests.cs

```diff
@@ -0,0 +1,308 @@
+                    (ReadResult.PropertyName, State.PendingPropertyNameRead),
+                    (ReadResult.PropertyValue, State.PendingPropertyValueRead),
+                    (ReadResult.Done, State.Done)
+                }
+            },
+            {
+                new[] { "{", """ "foo": """, """ "ba """, """ r" """, "}" },
+                new[]
+                {
+                    (ReadResult.Incomplete, State.PropertyNameOrEnd),
+                    (ReadResult.PropertyName, State.PendingPropertyNameRead),
+                    (ReadResult.PropertyValue, State.PendingPropertyValueRead),
+                    (ReadResult.PropertyValue, State.PendingPropertyValueRead),
+                    (ReadResult.Done, State.Done)
+                }
+            },
+            {
+                new[] { "{", "\"", "f", "o", "o", "\"", ":", "4","2", "}" },
+                new[]
+                {
+                    (ReadResult.Incomplete, State.PropertyNameOrEnd),
+                    (ReadResult.Incomplete, State.PropertyNameOrEnd),
+                    (ReadResult.Incomplete, State.PropertyNameOrEnd),
+                    (ReadResult.Incomplete, State.PropertyNameOrEnd),
+                    (ReadResult.Incomplete, State.PropertyNameOrEnd),
+                    (ReadResult.Incomplete, State.PropertyNameOrEnd),
+                    (ReadResult.PropertyName, State.PendingPropertyNameRead),
+                    (ReadResult.PropertyValue, State.PendingPropertyValueRead),
+                    (ReadResult.PropertyValue, State.PendingPropertyValueRead),
+                    (ReadResult.PropertyValue, State.PendingPropertyValueRead),
+                    (ReadResult.Done, State.Done)
+                }
+            },
+            {
+                new[] { "{", """ "foo": """, "123", "}" },
+                new[]
+                {
+                    (ReadResult.Incomplete, State.PropertyNameOrEnd),
+                    (ReadResult.PropertyName, State.PendingPropertyNameRead),
+                    (ReadResult.PropertyValue, State.PendingPropertyValueRead),
+                    (ReadResult.PropertyValue, State.PendingPropertyValueRead),
+                    (ReadResult.Done, State.Done)
+                }
+            },
+            {
+                new[] { "{", """ "foo":42 """, ",", """ "bar": "baz" """, "}" },
+                new[]
+                {
+                    (ReadResult.Incomplete, State.PropertyNameOrEnd),
+                    (ReadResult.PropertyName, State.PendingPropertyNameRead),
+                    (ReadResult.PropertyValue, State.PendingPropertyValueRead),
+                    (ReadResult.PropertyName, State.PendingPropertyNameRead),
+                    (ReadResult.PropertyValue, State.PendingPropertyValueRead),
+                    (ReadResult.Done, State.Done)
+                }
+            },
+            {
+                new[] { """ { "foo": """, /*lang=json*/ """ {"bar":42} """, "}" },
```

Comment:
nit: unless space is intentional?

```suggestion
                new[] { """ {"foo": """, /*lang=json*/ """ {"bar":42} """, "}" },
```